### ISSUE TITLE
Which description is true in GRaSP algorithm? the behavior seems that the required edges are respected.

### ISSUE THREAD

**Issue by yasu-sh** [cmu-phil/tetrad#1670]
Dear developers,

In index.html as manual and description(https://github.com/cmu-phil/tetrad/blob/development/docs/manual/index.html)

The description is written(including HTML page and TETRAD GUI) as below.
> Knowledge of forbidden edges may be used with GRaSP; currently knowledge of required edges is not implemented.

https://github.com/cmu-phil/tetrad/blame/656cee79e3a23c591a51f8a72aaf190476ba0bbb/tetrad-lib/src/main/java/edu/cmu/tetrad/search/Grasp.java#L45

But, in the class of Grasp, the description is opposite.
https://github.com/cmu-phil/tetrad/blob/development/tetrad-lib/src/main/java/edu/cmu/tetrad/search/Grasp.java

> This class is configured to respect knowledge of forbidden and required
>  edges, including knowledge of temporal tiers.</p>

https://github.com/cmu-phil/tetrad/blob/656cee79e3a23c591a51f8a72aaf190476ba0bbb/tetrad-lib/src/main/java/edu/cmu/tetrad/search/Grasp.java#L49C7-L49C7

I tried several settings. I guess the class description(bottom) is correct. I would appreciate if you could tell me and fix it.

**Comment by jdramsey** [comment]
Oh, thanks. The first one needs to be edited. Bryan has figured out how to do both required and forbidden edges with GRaSP, and it's currently implemented that way.

**Comment by yasu-sh** [comment]
@jdramsey Thanks very much for your comment! It makes clear my concern. Let me know a little more about the details.

My understanding is like this(I have not checked the code but I checked at the different conditions):
1. Search by Grow-Shrink starts with the knowledge of required edges and follows the forbidden edges rules including tiers.
2. After the initial search ended, the tucks manipulations start for searching better scores. but required edges in knowledg may disappear since "tuck" exchanges nodes as the arxiv paper(Wayne).

Is this correct or not? When I set 4 required edges in simulation box in TETRAD 7.5.0, 2 survived and 2 disappears.
So I wanted to confirm this.

**Comment by jdramsey** [comment]
@yasu-sh The strategy for incorporating knowledge for GRaSP is a little different because it's using permutations to infer graphs--i.e., it's putting variables in particular orders and inferring the DAGs from those variable orders. So the first thing it does is to try to find a permutation of the variables consistent with all of the knowledge (forbidden and required). Then it tries to make adjustments to that order that are also consistent with knowledge. Knowledge rules are also incorporated into the methods that infer the DAGs from the permutations to ensure that knowledge is satisfied at that level as well, which is helpful, for instance, if you forbid all edges within a tier, something you may not be able to accomplish by just putting the variables in a certain order and inferring a DAG--you may inadvertently that way end up with edges within a tier. So it's a little different from how, say, PC does it, but the result is that the search and the final graph are consistent with the background knowledge you provide.

**Comment by yasu-sh** [comment]
@jdramsey Thanks for your detailed explanation. It makes clearer understanding for knowledge on GRASP.
I guess the forbidden edges works completely on your explanation(not tried yet), I am not sure for required edges.

The reason is as below:
I should have sent you a screenshot for your understanding the sympton I experienced. I attached .tet files also.
At the flow of top side, x1 -> x8 and x5 -> x8 disappear.
For bottom side, x1 -> x8 and x1 -> x4 disappear.

![image](https://github.com/cmu-phil/tetrad/assets/89838961/d8e84533-4a56-49e7-a8d3-31cf9b390520)
tetrad.tet file for 7.5.0
[tetrad-issue1670.zip](https://github.com/cmu-phil/tetrad/files/12403360/tetrad-issue1670.zip)

If I should make some mistakes or misunderstandings, just let me know.
